File: webapp/Locust/Scripts/app/services/TestProcedureService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('TestProcedureService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/TestProcedure/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                Update: {
                    method: 'PUT'
                },
                Restore: {
                    method: 'POST'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Enable: {
                    method: 'GET',
                    isArray: false
                },
                GetLastTestProcedure: {
                    method: 'GET',
                    isArray: false
                }
                ,
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetLastOne: {
                    method: 'GET',
                    isArray: false
                },

                GetAll: {
                    method: 'GET',
                    isArray: true
                },

                UpdateNumber: {
                    method: 'PUT'
                },

                GetProject: {
                    method: 'GET',
                    isArray: false
                },
                GetRequirement: {
                    method: 'GET',
                    isArray: false
                },
                GetPtojectRequirement: {
                    method: 'GET',
                    isArray: false
                },
                AddChangeLog: {
                    method: 'GET',
                    isArray: false
                },
                TestProcedureChangeLogs: {
                    method: 'GET',
                    isArray: true
                },
                IsAssigned: {
                    method: 'GET',
                    isArray: false
                },
                GetRequirementForTp: {
                    method: 'GET',
                    isArray: false
                },
                CreateFromScripts: {
                    method: 'POST',
                    isArray: true
                }
            }),

            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };

        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };
        self.GetLastOne = function (idReq, creator, date, success, error) {
            return resource.GetLastOne({ action: 'GetLastOne', idReq: idReq, creator: creator, date: date }, success || noop, error || noop);
        };
        self.GetRequirement = function (reqId, success, error) {
            return resource.GetRequirement({ action: 'GetRequirement', reqId: reqId }, success || noop, error || noop);
        };
        self.Delete = function (id, success, error) {
            return resource.Delete({ action: 'Delete', id: id }, success || noop, error || noop);
        };
        self.Enable = function (id, success, error) {
            return resource.Enable({ action: 'Enable', id: id }, success || noop, error || noop);
        };
        self.GetLastTestProcedure = function (creator, date, success, error) {
            return resource.GetLastTestCase({ action: 'GetLastTestProcedure', creator: creator, date: date }, success || noop, error || noop)
        };
        self.GetProject = function (idtp, success, error) {
            return resource.GetProject({ action: 'GetProject', idtp: idtp }, success || noop, error || noop)
        };
        self.AddChangeLog = function (id, success, error) {
            return resource.AddChangeLog({ action: 'AddChangeLog', id: id }, success || noop, error || noop);
        };
        self.Restore = function (entity, success, error) {
            return resource.Restore({ action: 'Restore' }, entity, success || noop, error || noop);
        };
        self.GetProjectRequirement = function (idreq, success, error) {
            return resource.GetProject({ action: 'GetProjectRequirement', idreq: idreq }, success || noop, error || noop)
        };
        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };
        self.TestProcedureChangeLogs = function (id, success, error) {
            return resource.TestProcedureChangeLogs({ action: 'TestProcedureChangeLogs', id: id }, success || noop, error || noop);
        };
        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };
        self.UpdateNumber = function (entity, idReq, success, error) {
            return resource.UpdateNumber({ action: 'UpdateNumber', idReq: idReq }, entity, success || noop, error || noop);
        }
        self.IsAssigned = function (tpid, tcid, success, error) {
            return resource.IsAssigned({ action: 'IsAssigned',tpid, tcid: tcid }, success || noop, error || noop);
        }
        self.GetRequirementForTp = function (tpid, success, error) {
            return resource.GetRequirementForTp({ action: 'GetRequirementForTp', tpid: tpid }, success || noop, error || noop);
        }
        self.CreateFromScripts = function (scripts, projectId, success, error) {
            return resource.CreateFromScripts({ action: 'CreateFromScripts', projectId: projectId }, scripts, success || noop, error || noop);
        }
    }]);
})(angular);

File: webapp/Locus.Core/Services/TestProcedureService.cs
using Locus.Core.DTO;
using Locus.Core.Helpers;
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface ITestProcedureService
    {
        TestProcedure Save(TestProcedure testProcedure, string user);

        TestProcedure Update(TestProcedure testProcedure, string user);

        TestProcedure Delete(int idTestProcedure);

        TestProcedure Enable(int idTestProcedure);

        List<TestProcedure> GetAll();

        TestProcedure Get(int idTestProcedure);

        TestProcedure GetLastOne(int idReq, string creator, string date);

        TestProcedure GetLastTestProcedure(string creator, string date);

        TestProcedure UpdateNumber(TestProcedure testProcedure, int idReq);

        Project GetProject(int idtp);

        Requirement GetRequirement(int idtp);

        Project GetProjectRequirement(int idreq);

        ChangeLog AddChangeLog(int id, string user);

        List<ChangeLog> TestProcedureChangeLogs(int id);

        ChangeLog Restore(ChangeLog change_log, string user);

        TestProcedure GetTestProcedure(int test_Case_Id);

        AssignedStatusDTO IsAssigned(int? tpid,int? tcid);

        Requirement GetRequirementForTp(int tpId);
        List<TestProcedure> CreateFromScripts(Scripts[] scripts, int projectId, string user);
    }


    public class TestProcedureService : ITestProcedureService
    {
        private readonly ITestProcedureRepository _testProcedureRepository;
        private readonly IRequirementRepository _requirementRepository;
        private readonly ITestResultRepository _testResultRepository;

        //Constructor of TestCaseService and initialize testCaseRepository
        public TestProcedureService(ITestProcedureRepository testProcedureRepository)
        {
            _testProcedureRepository = testProcedureRepository;
        }

        public TestProcedureService(ITestProcedureRepository testProcedureRepository, ITestResultRepository testResultRepository, IRequirementRepository requirementRepository)
        {
            _testProcedureRepository = testProcedureRepository;
            _testResultRepository = testResultRepository;
            _requirementRepository = requirementRepository;
        }    

        public TestProcedure Delete(int idTestProcedure)
        {
            return _testProcedureRepository.Delete(idTestProcedure);
        }

        public TestProcedure Enable(int idTestProcedure)
        {
            return _testProcedureRepository.Enable(idTestProcedure);
        }

        public TestProcedure Get(int idTestProcedure)
        {
            return _testProcedureRepository.Get(idTestProcedure);
        }

        public List<TestProcedure> GetAll()
        {
            return _testProcedureRepository.GetAll();
        }

        public ChangeLog AddChangeLog(int testprocedureId, string user)
        {
            return _testProcedureRepository.AddChangeLog(testprocedureId, user);
        }

        public TestProcedure GetLastOne(int idReq, string creator, string date)
        {
            return _testProcedureRepository.GetLastOne(idReq, creator, date);
        }

        public TestProcedure GetLastTestProcedure(string creator, string date)
        {
            return _testProcedureRepository.GetLastTestProcedure(creator, date);
        }

        public Project GetProject(int idtp)
        {
            return _testProcedureRepository.GetProject(idtp);
        }

        public Project GetProjectRequirement(int idreq)
        {
            return _testProcedureRepository.GetProjectRequirement(idreq);
        }

        public Requirement GetRequirement(int idtp)
        {
            return _testProcedureRepository.GetRequirement(idtp);
        }

        public TestProcedure Save(TestProcedure testProcedure, string user)
        {
            return _testProcedureRepository.Save(testProcedure, user);
        }

        public TestProcedure Update(TestProcedure testProcedure, string user)
        {
            _testResultRepository.UpdateTitles(testProcedure.Test_Procedure_Id, testProcedure.Title, "TP");
            return _testProcedureRepository.Update(testProcedure, user);
        }

        public TestProcedure UpdateNumber(TestProcedure testProcedure, int idReq)
        {
            return _testProcedureRepository.UpdateNumber(testProcedure, idReq);
        }

        public List<ChangeLog> TestProcedureChangeLogs(int id)
        {
            return _testProcedureRepository.TestProcedureChangeLogs(id);
        }

        public ChangeLog Restore(ChangeLog change_log, string user)
        {
            return _testProcedureRepository.Restore(change_log, user);
        }

        public TestProcedure GetTestProcedure(int test_Case_Id)
        {
            return _testProcedureRepository.GetTestProcedureTc(test_Case_Id);
        }

        public AssignedStatusDTO IsAssigned(int? tpid ,int? tcid)
        {
            AssignedStatusDTO test = new AssignedStatusDTO();
            if (_testProcedureRepository.IsAssigned(tpid,tcid)) {
                test.assigned = true;
                test.message = "The Selected Test Case has been already been assigned to another Test Procedure";
                return test;
            }
            else
            {
                test.assigned = false;
                test.message = "";
                return test;
            }
            
        }

        public Requirement GetRequirementForTp(int tpId)
        {
            return _testProcedureRepository.GetRequirementForTp(tpId);
        }

        public List<TestProcedure> CreateFromScripts(Scripts[] scripts, int projectId, string user)
        {

            List<ScriptDTO> scriptDTOs = new List<ScriptDTO>();
            List<TestDTO> testCases = new List<TestDTO>();
            foreach (var script in scripts)
            {
                ScriptDTO scriptDTO = new ScriptDTO();
                scriptDTO.Id = script.Id;
                scriptDTO.RelatedTestCases = TextExtractorHelper.ExtractFromScript(script);
                 testCases.AddRange(_requirementRepository.GetTestCasesByNumber(scriptDTO.RelatedTestCases, projectId, script.Id));

            }

            return _testProcedureRepository.CreateTestProcedureFromTestCase(projectId, testCases, user);
        }
    }
}
